File: JavaScript/MaxHappinessSum.js
/**
 * @param {number[]} happiness
 * @param {number} k
 * @return {number}
 */
var maximumHappinessSum = function(happiness, k) {
  let ans = 0;
  
  happiness.sort((a, b) => b - a);
  
  for (let i = 0; i < k; i++){
    if (happiness[i] < 0)
      break;
    
    ans += happiness[i];
    
    if(i < happiness.length - 1)
      happiness[i + 1] -= (i + 1);
  }
  
  return ans;
};

File: Rust/MaxHappinessSum.rs
impl Solution {
    pub fn maximum_happiness_sum(mut happiness: Vec<i32>, k: i32) -> i64 {
        let mut ans = 0;
        
        happiness.sort_by(|a,b| b.cmp(a));
        
        for i in 0..k as usize {
            if happiness[i] < 0{
                break;
            }
            
            ans += happiness[i] as i64;
            
            if i < happiness.len() -1{
                happiness[i+1] -= (i + 1) as i32;
            }
        }
        
        ans
    }
}
